### PR TITLE
Add Support for Returning Logits for Generated Tokens

### LINKED ISSUE #588: Restore ability to get logits from generation?
It's often useful when generating samples to be able to get the logits / probabilities of the generated tokens (e.g. for ranking suggestions). It looks like this used to be available but was removed by 7aed13305c48b0048ec8b9efdba01e40fc69f123. It would be great if this functionality could be ported to the most recent version of the code.

I attempted to hack it back in, but must have messed something up because it made evaluation ~3.5x slower; I assume I accidentally introduced some GPU-CPU copies but unfortunately I don't really understand the Megatron/Deepspeed code well enough to see how to fix it. That branch is here: https://github.com/moyix/gpt-neox/tree/return_logits

Edit: fixed the link to my attempt at implementation, not sure how that happened

### PR DESCRIPTION
**Pull request by Kyle1668**
Closes #588

### CODE REVIEW

**Comment by Kyle1668** [comment]
The original issue mentioned potentially returning the log-likelihood of the tokens through some integration in the `generate.py` interface. I decided against implementing that support in this PR and instead returned the raw logits. My reasoning is that end users can then operate on these logits as needed.

**Comment by Kyle1668** [comment]
Outputs from ` ./deepy.py generate.py ./configs/20B.yml ./configs/text_generation.yml` on my branch

With `return_logits: true` 

image

without 

```json
{
  "context": "",
  "text": "Entry DescriptionWild Agnes has released this extraordinary new",
  "length": 10,
  "finished": false,
  "message": null,
  "duration_seconds": 6.626634120941162
}
```

on main

```json
{
  "context": "",
  "text": "Entry DescriptionWild Agnes has released this extraordinary new",
  "length": 10,
  "finished": false,
  "message": null,
  "duration_seconds": 6.523743629455566
}
```

**Comment by Kyle1668** [comment]
I took a stab at writing some unit tests for this change but ran into an issue with `mpu.get_model_parallel_src_rank()` throwing an exception saying that I needed to configure `init_process_group`.

**Review by StellaAthena — APPROVED** [review]
This is great! Apologies that this PR took so long to review and merge, but we hugely appreciate this work.
